code/obj/submachine/seed.dm

```diff
@@ -1,632 +1,494 @@
-			<h3 style='margin: 0;'>[src.name]</h3>
-			<div style="float: right;">
-				[src.inserted ? "<a href='?src=\ref[src];ejectbeaker=1' class='buttonlink'>&#9167;</a> [src.inserted] ([src.inserted.reagents.total_volume]/[src.inserted.reagents.maximum_volume]) &bull; " : "" ]
-				[src.extractables.len > 0 ? "<a href='?src=\ref[src];ejectextractables=1' class='buttonlink'>&#9167;</a> " : "" ][src.extractables.len] extractable\s &bull;
-				[src.seeds.len > 0 ? "<a href='?src=\ref[src];ejectseeds=1' class='buttonlink'>&#9167;</a> " : "" ][src.seeds.len] seed\s
-			</div>
-			<strong><a href='?src=\ref[src];page=1'>Overview</a> &bull; <a href='?src=\ref[src];page=2'>Seed Extraction</a> &bull; <a href='?src=\ref[src];page=3'>Seed List</a></strong>
-			<hr>
-		"}
-		if (src.mode == "overview")
-			dat += "<b><u>Overview</u></b><br><br>"
-
-			if (src.inserted)
-				dat += "<B>Receptacle:</B> [src.inserted] ([src.inserted.reagents.total_volume]/[src.inserted.reagents.maximum_volume]) <A href='?src=\ref[src];ejectbeaker=1'>(Eject)</A><BR>"
-				dat += "<b>Contents:</b> "
-				if(src.inserted.reagents.reagent_list.len)
-					for(var/current_id in inserted.reagents.reagent_list)
-						var/datum/reagent/current_reagent = inserted.reagents.reagent_list[current_id]
-						dat += "<BR><i>[current_reagent.volume] units of [current_reagent.name]</i>"
-				else
-					dat += "Empty"
-			else
-				dat += "<B>No receptacle inserted!</B>"
-
-			dat += "<br>"
-
-			if(src.seeds.len)
-				dat += "<BR><B>[src.seeds.len] Seeds Ready for Experimentation</B>"
-			else
-				dat += "<BR><B>No Seeds inserted!</B>"
-
-			dat += "<br>"
-
-			if(src.extractables.len)
-				dat += "<BR><B>[src.extractables.len] Items Ready for Extraction</B>"
-			else
-				dat += "<BR><B>No Extractable Produce inserted!</B>"
-
-		else if (src.mode == "extraction")
-			dat += "<b><u>Seed Extraction</u></b><br>"
-			if (src.seedoutput)
-				dat += "<A href='?src=\ref[src];outputmode=1'>Extracted seeds will be ejected from the machine.</A>"
-			else
-				dat += "<A href='?src=\ref[src];outputmode=1'>Extracted seeds will be retained within the machine.</A>"
-			dat += {"<br><br>
-				<table>
-					<thead>
-					<tr>
-						<th colspan="2">Name</th>
-						<th colspan='1'>Controls</th>
-						[genes_header]
-					</tr>
-					</thead>
-					<tbody>
-				"}
-
-			if(src.extractables.len)
-				for (var/obj/item/I in src.extractables)
-					var/geneout = ""
-					if (istype(I, /obj/item/seed))
-						var/obj/item/seed/S = I
+	ui_data(mob/user)
```

Comment:
can this not all be static data and then just manually updated whenever state changes?
really don't think we need to be running all of this every second